### PR TITLE
feat: Improve Performance at Migration

### PR DESCRIPTION
**Pull request by Aclrian**
Closes Safe-DS/Library-Analyzer#25.

### Summary of Changes
Improved Performance of the migrate command:
- add UnchangedDiffer that maps unchanged api-elements to its newer/older version.
- rewritten weighted levenshtein distance
- improve performance of StrictDiffer and its way to check if the parents of two api-elements are mapped to each other.

and some smaller changes:
- remove documentation before calculating the similarity of two codes
- move annotations that are migrated by ManyToManyMappings to the unsure annotation store

### Testing Instructions

migrate a package with a lot of api-elements and measure the time of the migration.

### CODE REVIEW

**Comment by lars-reimann** [comment]
## [🦙 MegaLinter](https://megalinter.io/6.19.0) status: ✅ [SUCCESS](https://github.com/Safe-DS/API-Editor/actions/runs/4414612515)

| Descriptor  |                                           Linter                                           |Files|Fixed|Errors|Elapsed time|
|-------------|--------------------------------------------------------------------------------------------|-----|----:|------|------------|
|✅ JSON      |[eslint-plugin-jsonc](https://megalinter.io/6.19.0/descriptors/json_eslint_plugin_jsonc)    |    7|    0|     0|16.6s       |
|✅ JSON      |[jsonlint](https://megalinter.io/6.19.0/descriptors/json_jsonlint)                          |    7|     |     0|0.63s       |
|✅ JSON      |[npm-package-json-lint](https://megalinter.io/6.19.0/descriptors/json_npm_package_json_lint)|yes  |     |no    |0.47s       |
|✅ JSON      |[prettier](https://megalinter.io/6.19.0/descriptors/json_prettier)                          |    7|    3|     0|12.92s      |
|✅ JSON      |[v8r](https://megalinter.io/6.19.0/descriptors/json_v8r)                                    |    7|     |     0|2.31s       |
|✅ PYTHON    |[black](https://megalinter.io/6.19.0/descriptors/python_black)                              |   15|    1|     0|2.39s       |
|✅ PYTHON    |[flake8](https://megalinter.io/6.19.0/descriptors/python_flake8)                            |   15|     |     0|0.85s       |
|✅ PYTHON    |[isort](https://megalinter.io/6.19.0/descriptors/python_isort)                              |   15|    1|     0|1.61s       |
|✅ PYTHON    |[mypy](https://megalinter.io/6.19.0/descriptors/python_mypy)                                |   15|     |     0|3.24s       |
|✅ PYTHON    |[pylint](https://megalinter.io/6.19.0/descriptors/python_pylint)                            |   15|     |     0|5.64s       |
|✅ REPOSITORY|[git_diff](https://megalinter.io/6.19.0/descriptors/repository_git_diff)                    |yes  |     |no    |0.73s       |

See detailed report in [MegaLinter reports](https://github.com/Safe-DS/API-Editor/actions/runs/4414612515)
_Set `VALIDATE_ALL_CODEBASE: true` in mega-linter.yml to validate all sources, not only the diff_

_MegaLinter is graciously provided by [OX Security](https://www.ox.security/?ref=megalinter)_

**Comment by Aclrian** [comment]
grafik

File: package-parser/package_parser/processing/api/model/_api.py [hunk 1]
```diff
@@ -292,14 +302,49 @@ def to_json(self) -> Any:
...
+            and self.decorators == other.decorators
+            and self.superclasses == other.superclasses
+            and self.methods == other.methods
+            and self.is_public == other.is_public
+            and self.reexported_by == other.reexported_by
+            and self.documentation == other.documentation
+            and self.code == other.code
+            and self.instance_attributes == other.instance_attributes
+        )
+
```
Comment (lars-reimann) [inline package-parser/package_parser/processing/api/model/_api.py:347]: `__hash__` and `__eq__` already get created for a dataclass. Is there a particular reason you have to override them?

**Comment by lars-reimann** [comment]
@Aclrian Since this repo is about to get more busy very soon, I could not wait to finish #1258 any longer. The former `package-parser` is now called `library-analyzer` and resides in its own repo: https://github.com/Safe-DS/Library-Analyzer. Once you have the time, kindly

1. Clone https://github.com/Safe-DS/Library-Analyzer
2. Create a new branch there
3. Copy the files you changed here over there
 * `src` contains the sources
 * `tests` contains the tests
 * `package_parser` is `library_analyzer`
 * Otherwise, the structure is identical so far
4. Create a PR (with the same description as here)

Sorry for the disruption and inconvenience.

@Masara, @lukarade, @nvollroth Please do the same if you already created branches with changes to the former `package-parser`.

File: package-parser/package_parser/processing/api/model/_api.py (hunk 1 above)
Comment (Aclrian) [reply package-parser/package_parser/processing/api/model/_api.py:347]: For parameters: The necessary frozen attribute that is needed for the creation of a hash function would conflict with the creation of types in the init function if frozen is set True.
For all other cases, `__eq__` and `__hash__` will be removed and frozen will be set True for Safe-DS/Library-Analyzer#25

**Comment by Aclrian** [comment]
will be completed in Safe-DS/Library-Analyzer#62.

**Pull request closed without merging**